### ISSUE TITLE
JUnit Platform aka JUnit 5 Support

### ISSUE THREAD

**Issue by eriwen** [gradle/gradle#828]
Provide built-in support for JUnit Platform. 

### Context
Discussion with [Marc from JUnit team](https://discuss.gradle.org/t/core-support-for-junit-platform-a-k-a-junit-5/19487)

Currently the JUnit maintains [a Gradle plugin](https://github.com/junit-team/junit5/tree/master/junit-platform-gradle-plugin/src/main/groovy/org/junit/platform/gradle/plugin), but we would like to see this in Gradle.

### Additional Use Cases
Consider gradle/gradle#701 regarding allowing custom implementation of the bootstrap routine.

**Comment by ajoberstar** [comment]
If you want any community help on this one, I'd be interested. While working on some testing support for Clojure with JUnit Platform, I took an initial pass at integrating it into Gradle in [jovial](https://github.com/ajoberstar/jovial). Got very hacky due to limitations of the current Test infrastructure.

As much as I believe JUnit Platform is a great integration point for most testing frameworks and definitely want it to have core support, I think the Test infrastructure should be decoupled from any specific framework. Most of the challenges when I worked on this were due to assumptions that no one would ever want to integrate an alternate test framework into Gradle, which holds you back from a lot of key features, like integration into Tooling API, and other eventing mechanisms.

**Comment by eriwen** [comment]
That would be awesome @ajoberstar! I agree regarding test framework abstraction. 

The first step is to write a design spec. There is some useful information on the linked forum topic. 

We are willing to provide significant help with design and development, though we won't be able to put in much more than a few hours until Gradle 3.3 is feature complete in December.

What level of involvement are you comfortable with in the design phase?

**Comment by ajoberstar** [comment]
Sounds good. I'll take a crack at a design doc in the next week or so, and we can see if it's in the ballpark of what you're looking for. If it needs to wait until after December at that point, no big deal from my perspective.

**Comment by wlsc** [comment]
Any news on this issue?

**Comment by eriwen** [comment]
@wlsc We have met with the JUnit team and have done a spike and determined that Gradle has all necessary APIs for initial support. @ajoberstar is helping us out to spike integration with Gradle, and once that's done we will discuss our options. 

Forward progress being made, but I would not expect built-in support until Gradle 4.6 at the earliest.

**Comment by oehme** [comment]
Just a quick update: Thanks to @ajoberstar's spikes and a discussion with the JUnit team we have a good plan for how to go forward. @blindpirate is currently working on the [first step](https://github.com/gradle/gradle/pull/4116) of this plan.

1. Our first iteration will do no filtering of classes, but just pass everything to JUnit. We will use the usual round-robin algorithm to assign tests to worker processes. We will execute all tests of a worker in one batch so that extensions are only initialized once. We’ll map the results back to the existing JUnit4 style report.
2. The next most important feature is a DSL that allows configuration of the test plan (e.g. includes and excludes)
3. After that we’ll revisit our TestListener so it covers enough information to uniquely identify a test (currently it only has a name and class name). This should also be exposed to the Tooling API
4. We might want to reconsider our parallelisation approach, maybe using JUnit’s discovery up-front to avoid scheduling ignored tests and better balancing tests over workers
5. The JUnit team is working on a better report format. We’ll support this format once it is ready.
6. Some frameworks like cucumber are not class-based - We’ll need to consider how to support those, e.g. allowing resources as inputs for test discovery. Let’s wait till cucumber is ready and then use it as an example to drive the design.

**Comment by JLLeitschuh** [comment]
For number 1:
>  We will use the usual round-robin algorithm to assign tests to worker processes. We will execute all tests of a worker in one batch so that extensions are only initialized once. 

How will you support tests that require static state (not a great way to write tests but its a use case that needs to be considered). For example, a webserver to be spun up. If Junit 5 tests are being split into multiple workers, do these workers have isolated classloaders? Will there be a way of specifying that all tests should be run in the same worker?

Please excuse me if I'm totally misunderstanding something here.

**Comment by oehme** [comment]
@JLLeitschuh A worker is a separate process, so completely isolated. The number of workers per test task is controlled by `Test.maxParallelForks`.

**Comment by Thunderforge** [comment]
For those who didn't spot it, [Gradle 4.6 RC1](https://github.com/gradle/gradle/releases/tag/v4.6.0-RC1) adds JUnit 5 support.

> Next, this release of Gradle includes built-in support for JUnit Platform and the JUnit Jupiter/Vintage Engine, also known as [JUnit 5 support](https://docs.gradle.org/4.6-rc-1/release-notes.html#junit-5-support). You can use the new filtering and engines functionality in JUnit 5 using the examples provided below and in the documentation.

[Gradle 4.6 RC2](https://github.com/gradle/gradle/releases/tag/v4.6.0-RC2) adds additional documentation, per #4403.

**Comment by ajoberstar** [comment]
@oehme Are those steps from this new plan opened as issues? For [gradle-clojure](https://github.com/gradle-clojure/gradle-clojure), I'd prefer to switch to JUnit Platform based test running, as the JUnit 4 option gets a little gross. I think I'll need step 6 for that to work out cleanly, and would like to track the progress you guys make towards that.

**Comment by oehme** [comment]
@ajoberstar I just created https://github.com/gradle/gradle/issues/4773

**Comment by ajoberstar** [comment]
@oehme Thanks

**Comment by childnode** [comment]
@oehme / @eriwen I (and I bet some more) would be pleased if the open tasks / plans in your TOP comments of this epic can be marked as done and open ones can be marked / annotated to what is missing etc. to give us a quick overview

**Comment by oehme** [comment]
@childnode If you install [ZenHub](https://www.zenhub.com/) you'll see that this issue is actually an epic and has other issues contained in it, some of which are complete and some of which are open.

**Comment by ajoberstar** [comment]
Looks like you can get there without installing ZenHub too.

https://app.zenhub.com/workspace/o/gradle/gradle

**Comment by christopher-johnson** [comment]
Any plan for supporting testing in JPMS?  Not sure if you know this, but enabling the useJUnitPlatform configuration in a JDK 9+ module build causes ClassCastException in tests for anything that uses a Logger.

**Comment by blindpirate** [comment]
@christopher-johnson No I'm not aware of that. Could you please create a new issue in https://github.com/gradle/gradle/issues/new ? A minimum reproducible sample would be great.

**Comment by christopher-johnson** [comment]
New Issue created https://github.com/gradle/gradle/issues/4974.  Test repo is https://github.com/christopher-johnson/gradle-junit5-jpms

**Comment by raniejade** [comment]
Is there any plans to remove the assumption on how tests are structured? e.g. That it's always in the form of:

```kotlin
class TestSuite{
   @Test
   fun testMethod() { ... }

   @Test
   fun anotherTestMethod() { ... }
}
```

Reporting is broken when using [Spek](https://github.com/spekframework/spek) and the native support for JUnit Platform.

**Comment by oehme** [comment]
Can you please open an issue with a reproducible example?

**Comment by raniejade** [comment]
@oehme #5273

**Comment by ljacomet** [comment]
Closing this as the support is available since Gradle 4.6.

While there may be missing features and bugs, it works for a large number of use cases.
